h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -3579,177 +3957,191 @@ private void doCompute() {
+                  : _model._output._submodels.length - 1;
+          alphaStart = submodelCount / _parms._lambda.length;
+          lambdaStart = submodelCount % _parms._lambda.length;
+        }
+        _model._output._lambda_array_size = _parms._lambda.length;
+        for (int alphaInd = alphaStart; alphaInd < _parms._alpha.length; alphaInd++) {
+          _state.setAlpha(_parms._alpha[alphaInd]);   // loop through the alphas
+          if ((!_parms._HGLM) && (alphaInd > 0) && !_checkPointFirstIter) // no need for cold start during the first iteration
+            coldStart(devHistoryTrain, devHistoryTest);  // reset beta, lambda, currGram
+          for (int i = lambdaStart; i < _parms._lambda.length; ++i) {  // for lambda search, can quit before it is done
+            if (_job.stop_requested() || (timeout() && _model._output._submodels.length > 0))
+              break;  //need at least one submodel on timeout to avoid issues.
+            if (_parms._max_iterations != -1 && _state._iter >= _parms._max_iterations)
+              break;  // iterations accumulate across all lambda/alpha values when coldstart = false
+            if ((!_parms._HGLM && (_parms._cold_start || (!_parms._lambda_search && _parms._cold_start))) && (i > 0)
+                    && !_checkPointFirstIter) // default: cold_start for non lambda_search
+              coldStart(devHistoryTrain, devHistoryTest);
+            Submodel sm = computeSubmodel(submodelCount, _parms._lambda[i], nullDevTrain, nullDevValid);
+            if (_checkPointFirstIter)
+              _checkPointFirstIter = false;
+            double trainDev = sm.devianceTrain; // this is stupid, they are always -1 except for lambda_search=True
+            double testDev = sm.devianceValid;
+            devHistoryTest[submodelCount % devHistoryTest.length] =
+                    (oldDevTest - testDev) / oldDevTest; // only remembers 5
+            oldDevTest = testDev;
+            devHistoryTrain[submodelCount % devHistoryTrain.length] =
+                    (oldDevTrain - trainDev) / oldDevTrain;
+            oldDevTrain = trainDev;
+            if (_parms._lambda[i] < _lmax && Double.isNaN(_lambdaCVEstimate) /** if we have cv lambda estimate we should use it, can not stop before reaching it */) {
+              if (_parms._early_stopping && _state._iter >= devHistoryTrain.length) {
+                double s = ArrayUtils.maxValue(devHistoryTrain);
+                if (s < 1e-4) {
+                  Log.info(LogMsg("converged at lambda[" + i + "] = " + _parms._lambda[i] + "alpha[" + alphaInd + "] = "
+                          + _parms._alpha[alphaInd] + ", improvement on train = " + s));
                   break; // started overfitting
                 }
+                if (_validDinfo != null && _parms._nfolds <= 1) { // check for early stopping on test with no xval
+                  s = ArrayUtils.maxValue(devHistoryTest);
```

Comment:
If there could be `NaN`s in `devHistoryTest` (which I think is possible), I'd be careful with using `ArrayUtils.maxValue`:

```java
 ArrayUtils.maxValue(new double[] {Double.NaN, 1.,2.3,4.}); // => NaN
 ArrayUtils.maxValue(new double[] {1.,Double.NaN, 1.,2.3,4.}); // => 4.0
 ```
 
 `maxValue` takes a first item and compares it with others, if the first thing is `NaN` then all the evaluations end up with `False`. 
 
 https://github.com/h2oai/h2o-3/blob/5768898453ba6bcfc5b4d12f5dc404b8be8573b4/h2o-core/src/main/java/water/util/ArrayUtils.java#L853-L858

Reply:
Fixed.